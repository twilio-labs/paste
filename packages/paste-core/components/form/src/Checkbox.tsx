import * as React from 'react';
import * as PropTypes from 'prop-types';
import {useUID} from 'react-uid';
import {styled} from '@twilio-paste/styling-library';
import {Box} from '@twilio-paste/box';
import {BackgroundColorOptions, SpaceOptions} from '@twilio-paste/style-props';
import {
  BaseRadioCheckboxControl,
  BaseRadioCheckboxLabel,
  BaseRadioCheckboxLabelText,
  BaseRadioCheckboxHelpText,
} from './shared/BaseRadioCheckbox';

/* eslint-disable emotion/syntax-preference */
const HiddenCheckbox = styled.input({
  border: '0',
  clip: 'rect(0 0 0 0)',
  height: '1px',
  margin: '-1px',
  overflow: 'hidden',
  padding: '0',
  position: 'absolute',
  textTransform: 'none',
  whiteSpace: 'nowrap',
  width: '1px',
});
/* eslint-enable */

export interface CheckboxProps extends React.InputHTMLAttributes<HTMLInputElement> {
  children: NonNullable<React.ReactNode>;
  hasError?: boolean;
  helpText?: string | React.ReactNode;
  id: string;
  indeterminate?: boolean;
  isSelectAll?: boolean;
  isSelectAllChild?: boolean;
}
const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(
  (
    {id, checked, children, helpText, disabled, hasError, indeterminate, isSelectAllChild, isSelectAll, ...props},
    ref
  ) => {
    const helpTextId = useUID();

    let checkboxBackground: BackgroundColorOptions | null = null;
    if (isSelectAll) checkboxBackground = 'colorBackground';
    if ((isSelectAll && checked) || (isSelectAll && indeterminate))
      checkboxBackground = 'colorBackgroundPrimaryLightest';

    let paddingLeft: SpaceOptions | null = isSelectAll ? 'space20' : null;
    if (isSelectAllChild) paddingLeft = 'space30';

    return (
      <Box
        backgroundColor={checkboxBackground}
        borderRadius={isSelectAll ? 'borderRadius10' : null}
        display="inline-flex"
        flexDirection="column"
        padding={isSelectAll ? 'space30' : null}
        paddingBottom={isSelectAll ? 'space20' : null}
        paddingLeft={paddingLeft}
        paddingRight={isSelectAllChild ? 'space30' : null}
        paddingTop={isSelectAll ? 'space20' : null}
      >
        <HiddenCheckbox
          {...props}
          aria-describedby={helpTextId}
          aria-checked={indeterminate ? 'mixed' : checked}
          aria-invalid={hasError}
          checked={checked}
          disabled={disabled}
          id={id}
          ref={ref}
          type="checkbox"
        />
        <BaseRadioCheckboxLabel disabled={disabled} htmlFor={id}>
          <BaseRadioCheckboxControl
            alignItems="center"
            borderRadius="borderRadius10"
            disabled={disabled}
            display="flex"
            position="relative"
            type="checkbox"
          >
            <Box
              alignItems="center"
              as="span"
              display="flex"
              height={!indeterminate ? 'sizeSquare25' : undefined}
              position="absolute"
              top={!indeterminate ? '3px' : undefined}
              transform={!indeterminate ? 'rotate(-45deg)' : undefined}
              width="sizeSquare30"
              _before={
                !indeterminate
                  ? {
                      borderColor: disabled && checked ? 'colorBorderDark' : 'colorBorderInverse',
                      borderRadius: 'borderRadius10',
                      borderStyle: 'solid',
                      borderWidth: 'borderWidth10',
                      bottom: '0',
                      content: '""',
                      display: 'block',
                      height: 'sizeSquare25',
                      position: 'absolute',
                      width: 'size0',
                    }
                  : undefined
              }
              _after={{
                borderColor: disabled && checked ? 'colorBorderDark' : 'colorBorderInverse',
                borderRadius: 'borderRadius10',
                borderStyle: 'solid',
                borderWidth: 'borderWidth10',
                bottom: '0',
                content: '""',
                display: 'block',
                height: 'size0',
                position: !indeterminate ? 'absolute' : undefined,
                width: 'sizeSquare30',
              }}
            />
          </BaseRadioCheckboxControl>
          <BaseRadioCheckboxLabelText fontWeight={isSelectAll ? null : 'fontWeightNormal'}>
            {children}
          </BaseRadioCheckboxLabelText>
        </BaseRadioCheckboxLabel>
        {helpText && <BaseRadioCheckboxHelpText helpTextId={helpTextId}>{helpText}</BaseRadioCheckboxHelpText>}
      </Box>
    );
  }
);
Checkbox.displayName = 'Checkbox';

if (process.env.NODE_ENV === 'development') {
  Checkbox.propTypes = {
    children: PropTypes.node.isRequired,
    hasError: PropTypes.bool,
    helpText: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    id: PropTypes.string.isRequired,
    indeterminate: PropTypes.bool,
    isSelectAll: PropTypes.bool,
    isSelectAllChild: PropTypes.bool,
    value: PropTypes.string.isRequired,
  };
}

export {Checkbox};
