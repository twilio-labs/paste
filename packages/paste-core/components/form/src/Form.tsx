import { Box, safelySpreadBoxProps } from "@twilio-paste/box";
import type { BoxProps, BoxStyleProps } from "@twilio-paste/box";
import * as React from "react";

export interface FormProps extends Omit<React.ComponentPropsWithRef<"form">, "children"> {
  element?: BoxProps["element"];
  children: React.ReactNode;
  maxWidth?: BoxStyleProps["maxWidth"];
}

export const Form = React.forwardRef<HTMLDivElement, FormProps>(
  ({ element = "FORM", maxWidth, children, ...props }, ref) => (
    <Box
      as="form"
      ref={ref}
      element={element}
      maxWidth={maxWidth}
      display="flex"
      flexDirection="column"
      rowGap="space80"
      {...safelySpreadBoxProps(props)}
    >
      {children}
    </Box>
  ),
);

Form.displayName = "Form";
