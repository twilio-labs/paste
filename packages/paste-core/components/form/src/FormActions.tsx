import * as React from 'react';
import PropTypes from 'prop-types';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import type {BoxProps} from '@twilio-paste/box';

export interface FormActionsProps extends Omit<React.ComponentPropsWithRef<'div'>, 'children'> {
  element?: BoxProps['element'];
  children: React.ReactNode;
}

export const FormActions = React.forwardRef<HTMLDivElement, FormActionsProps>(
  ({children, element = 'FORM_ACTIONS', ...props}, ref) => (
    <Box
      ref={ref}
      element={element}
      {...safelySpreadBoxProps(props)}
      display="flex"
      flexDirection="row"
      columnGap="space40"
      marginTop="space60"
    >
      {children}
    </Box>
  )
);

FormActions.displayName = 'FormActions';

FormActions.propTypes = {
  children: PropTypes.node,
  element: PropTypes.string,
};
