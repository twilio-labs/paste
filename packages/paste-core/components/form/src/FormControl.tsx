import * as React from 'react';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import type {BoxProps} from '@twilio-paste/box';

export interface FormControlProps extends Omit<React.ComponentPropsWithRef<'div'>, 'children'> {
  element?: BoxProps['element'];
  children: React.ReactNode;
}

export const FormControl = React.forwardRef<HTMLDivElement, FormControlProps>(
  ({children, element = 'FORM_CONTROL', ...props}, ref) => (
    <Box ref={ref} flexGrow={1} element={element} {...safelySpreadBoxProps(props)}>
      {children}
    </Box>
  )
);

FormControl.displayName = 'FormControl';
