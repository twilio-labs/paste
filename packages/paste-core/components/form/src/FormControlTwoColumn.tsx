import { Box, safelySpreadBoxProps } from "@twilio-paste/box";
import type { BoxProps } from "@twilio-paste/box";
import type { HTMLPasteProps } from "@twilio-paste/types";
import * as React from "react";

export interface FormControlTwoColumnProps extends HTMLPasteProps<"legend"> {
  /**
   * Overrides the default element name to apply unique styles with the Customization Provider
   *
   * @default 'FORM_CONTROL_TWO_COLUMN'
   * @type {BoxProps['element']}
   * @memberof FormControlTwoColumnProps
   */
  element?: BoxProps["element"];
  children: React.ReactNode;
}

export const FormControlTwoColumn = React.forwardRef<HTMLLegendElement, FormControlTwoColumnProps>(
  ({ children, element = "FORM_CONTROL_TWO_COLUMN", ...props }, ref) => (
    <Box
      display="grid"
      gridTemplateColumns="1fr 1fr"
      columnGap="space50"
      ref={ref}
      element={element}
      {...safelySpreadBoxProps(props)}
    >
      {children}
    </Box>
  ),
);

FormControlTwoColumn.displayName = "FormControlTwoColumn";
