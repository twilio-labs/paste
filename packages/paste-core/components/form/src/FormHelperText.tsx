import * as React from 'react';
import * as PropTypes from 'prop-types';
import {ValueOf} from '@twilio-paste/types';
import {Flex} from '@twilio-paste/flex';
import {Text, safelySpreadTextProps} from '@twilio-paste/text';
import {ErrorIcon} from '@twilio-paste/icons/esm/ErrorIcon';

export const FormHelperTextVariants = {
  ERROR: 'error',
} as const;

export type FormHelperTextVariants = ValueOf<typeof FormHelperTextVariants>;

export interface FormHelperTextProps extends React.HTMLAttributes<''> {
  marginTop?: 'space0';
  className?: never;
  style?: never;
  variant?: FormHelperTextVariants;
}

const FormHelperText: React.FC<FormHelperTextProps> = ({marginTop, children, variant, ...props}) => {
  let icon = null;
  switch (variant) {
    case FormHelperTextVariants.ERROR:
      icon = <ErrorIcon iconColor="colorTextError" decorative />;
      break;
    default:
      break;
  }

  return (
    <Flex vAlignContent="center" marginTop={marginTop || 'space20'}>
      {icon}
      <Text
        {...safelySpreadTextProps(props)}
        as="div"
        fontSize="fontSize30"
        lineHeight="lineHeight30"
        textColor={variant === 'error' ? 'colorTextError' : 'colorTextWeak'}
        marginLeft={icon ? 'space10' : undefined}
      >
        {children}
      </Text>
    </Flex>
  );
};

FormHelperText.displayName = 'FormHelperText';

FormHelperText.propTypes = {
  marginTop: PropTypes.oneOf(['space0']),
};

export {FormHelperText};
