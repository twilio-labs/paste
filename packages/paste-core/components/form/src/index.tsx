export * from './FormLabel';
export * from './FormInput';
export * from './FormHelpText';
export * from './FormTextArea';
export * from '@twilio-paste/checkbox';
export * from '@twilio-paste/radio-group';
export * from '@twilio-paste/select';
