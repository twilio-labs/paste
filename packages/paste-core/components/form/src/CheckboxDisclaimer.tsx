import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box} from '@twilio-paste/box';
import {Checkbox, CheckboxProps} from './Checkbox';
import {FormHelpText} from './FormHelpText';

export interface CheckboxDisclaimerProps
  extends Omit<CheckboxProps, 'isSelectAll' | 'isSelectAllChild' | 'indeterminate' | 'helpText' | 'hasError'> {
  children: NonNullable<React.ReactNode>;
  errorText?: string | React.ReactNode;
}

const CheckboxDisclaimer: React.FC<CheckboxDisclaimerProps> = ({children, errorText, ...props}) => {
  return (
    <>
      <Box backgroundColor="colorBackground" borderRadius="borderRadius20" padding="space50">
        <Checkbox {...props} hasError={errorText != null}>
          {children}
        </Checkbox>
      </Box>
      {errorText && (
        <Box marginTop="space30" marginLeft="space50">
          <FormHelpText variant="error">{errorText}</FormHelpText>
        </Box>
      )}
    </>
  );
};

CheckboxDisclaimer.displayName = 'CheckboxDisclaimer';

if (process.env.NODE_ENV === 'development') {
  CheckboxDisclaimer.propTypes = {
    children: PropTypes.node.isRequired,
    errorText: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  };
}

export {CheckboxDisclaimer};
