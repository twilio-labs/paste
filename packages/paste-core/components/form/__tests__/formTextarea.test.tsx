import * as React from 'react';
import {render} from 'react-dom';
import renderer from 'react-test-renderer';
import {shallow, ReactWrapper, mount} from 'enzyme';
import {axe} from 'jest-axe';
import {Theme} from '@twilio-paste/theme';
import {FormHelpText, FormLabel, FormTextArea} from '../src';

const NOOP = (): void => {};

describe('FormTextArea render', () => {
  it('it should render', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider theme="console">
          <FormTextArea id="textarea" onChange={NOOP} />
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('it has no accessibility violations', async () => {
    const container = document.createElement('div');
    document.body.append(container);
    render(
      <Theme.Provider theme="console">
        <FormLabel htmlFor="textarea_1">Label Text</FormLabel>
        <FormTextArea id="textarea_1" onChange={NOOP} />
      </Theme.Provider>,
      container
    );
    const results = await axe(document.body);
    expect(results).toHaveNoViolations();
  });

  it('disabled has no accessibility violations', async () => {
    const container = document.createElement('div');
    document.body.append(container);
    render(
      <Theme.Provider theme="console">
        <FormLabel htmlFor="textarea_2" disabled>
          Label Text
        </FormLabel>
        <FormTextArea id="textarea_2" onChange={NOOP} disabled />
      </Theme.Provider>,
      container
    );
    const results = await axe(document.body);
    expect(results).toHaveNoViolations();
  });

  it('hasError has no accessibility violations', async () => {
    const container = document.createElement('div');
    document.body.append(container);
    render(
      <Theme.Provider theme="console">
        <FormLabel htmlFor="textarea_3">Label Text</FormLabel>
        <FormTextArea id="textarea_3" onChange={NOOP} hasError />
        <FormHelpText variant="error">Error info. Explains why the input has an error.</FormHelpText>
      </Theme.Provider>,
      container
    );
    const results = await axe(document.body);
    expect(results).toHaveNoViolations();
  });
});

describe('FormTextArea readyOnly prop', () => {
  const initialProps = {
    id: 'textarea',
    onChange: NOOP,
    readOnly: true,
  };

  const container = shallow(<FormTextArea {...initialProps} />);

  it('should have a readOnly prop', () => {
    expect(container.find('FieldWrapper').prop('readOnly')).toEqual(true);
  });

  it('should have a aria-readonly prop', () => {
    expect(container.find('TextAreaElement').prop('aria-readonly')).toEqual(true);
  });

  it('should have a readonly prop on InputElement', () => {
    expect(container.find('TextAreaElement').prop('aria-readonly')).toEqual(true);
  });
});

describe('FormTextArea disabled prop', () => {
  const initialProps = {
    id: 'textarea',
    onChange: NOOP,
    disabled: true,
  };

  const container = shallow(<FormTextArea {...initialProps} />);

  it('should have a disabled prop', () => {
    expect(container.find('FieldWrapper').prop('disabled')).toEqual(true);
  });

  it('should have a disabled prop on InputElement', () => {
    expect(container.find('TextAreaElement').prop('disabled')).toEqual(true);
  });
});

describe('FormTextArea hasError prop', () => {
  const initialProps = {
    id: 'textarea',
    onChange: NOOP,
    hasError: true,
  };

  const container = shallow(<FormTextArea {...initialProps} />);

  it('should have a hasError prop', () => {
    expect(container.find('FieldWrapper').prop('hasError')).toEqual(true);
  });

  it('should have a aria-invalid prop', () => {
    expect(container.find('TextAreaElement').prop('aria-invalid')).toEqual(true);
  });
});

describe('FormTextArea id prop', () => {
  const initialProps = {
    id: 'textarea',
    onChange: NOOP,
  };

  const container = shallow(<FormTextArea {...initialProps} />);

  it('should have an id prop', () => {
    expect(container.find('TextAreaElement').prop('id')).toEqual('textarea');
  });
});

describe('FormTextArea name prop', () => {
  const initialProps = {
    id: 'textarea',
    name: 'name',
    onChange: NOOP,
  };

  const container = shallow(<FormTextArea {...initialProps} />);

  it('should have an name prop', () => {
    expect(container.find('TextAreaElement').prop('name')).toEqual('name');
  });
});

describe('FormTextArea placeholder prop', () => {
  const initialProps = {
    id: 'textarea',
    placeholder: 'placeholder',
    onChange: NOOP,
  };

  const container = shallow(<FormTextArea {...initialProps} />);

  it('should have an placeholder prop', () => {
    expect(container.find('TextAreaElement').prop('placeholder')).toEqual('placeholder');
  });
});

describe('FormTextArea insertBefore prop', () => {
  const initialProps = {
    id: 'textarea',
    insertBefore: '<div>$10.99</div>',
    onChange: NOOP,
  };

  const container = shallow(<FormTextArea {...initialProps} />);

  it('should have a Prefix', () => {
    expect(container.find('Prefix').length).toEqual(1);
  });
});

describe('FormTextArea insertAfter prop', () => {
  const initialProps = {
    id: 'textarea',
    insertAfter: '<div>$10.99</div>',
    onChange: NOOP,
  };

  const container = shallow(<FormTextArea {...initialProps} />);

  it('should have a Suffix', () => {
    expect(container.find('Suffix').length).toEqual(1);
  });
});

describe('FormTextArea event handlers', () => {
  it('Should call the appropriate event handlers', () => {
    const onChangeMock: jest.Mock = jest.fn();
    const onFocusMock: jest.Mock = jest.fn();
    const onBlurMock: jest.Mock = jest.fn();

    const wrapper: ReactWrapper = mount(
      <FormTextArea id="textarea" onChange={onChangeMock} onFocus={onFocusMock} onBlur={onBlurMock} />
    );

    wrapper.find('textarea').simulate('change');
    expect(onChangeMock).toHaveBeenCalledTimes(1);
    wrapper.find('textarea').simulate('focus');
    expect(onFocusMock).toHaveBeenCalledTimes(1);
    wrapper.find('textarea').simulate('blur');
    expect(onBlurMock).toHaveBeenCalledTimes(1);
  });
});
