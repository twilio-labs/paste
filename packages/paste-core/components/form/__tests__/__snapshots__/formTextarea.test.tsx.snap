// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTextArea render it should render 1`] = `
.emotion-3 {
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5rem;
  color: #282a2b;
  font-weight: 400;
}

.emotion-3 *,
.emotion-3 *::after,
.emotion-3 *::before {
  box-sizing: border-box;
}

.emotion-2 {
  box-sizing: border-box;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  background-color: #ffffff;
  box-shadow: 0 0 0 1px #606b85;
  border-radius: 4px;
  -webkit-transition: box-shadow 100ms ease-in;
  transition: box-shadow 100ms ease-in;
  cursor: text;
}

.emotion-2:hover {
  box-shadow: 0 0 0 1px #030b5d;
}

.emotion-2:focus-within {
  box-shadow: 0 0 0 4px rgba(2,122,197,0.8);
}

.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: none;
  background: transparent;
  outline: none;
  display: block;
  width: 100%;
  max-height: 18.5rem;
  font-size: 0.875rem;
  line-height: 1.5rem;
  font-weight: 400;
  color: #282a2b;
  padding-top: 0.5rem;
  padding-right: 0.75rem;
  padding-bottom: 0.5rem;
  padding-left: 0.75rem;
  border-radius: 4px;
  resize: vertical;
}

.emotion-0::-webkit-input-placeholder {
  color: #666a6d;
  font-style: italic;
}

.emotion-0::-moz-placeholder {
  color: #666a6d;
  font-style: italic;
}

.emotion-0:-ms-input-placeholder {
  color: #666a6d;
  font-style: italic;
}

.emotion-0::placeholder {
  color: #666a6d;
  font-style: italic;
}

.emotion-0:disabled {
  cursor: not-allowed;
  color: #666a6d;
}

<div
  className="emotion-3"
>
  <div
    className="emotion-2"
    cursor="text"
    display="flex"
    width="100%"
  >
    <textarea
      className="emotion-0 emotion-1"
      id="textarea"
      onChange={[Function]}
      rows={3}
      spellCheck={true}
    />
  </div>
</div>
`;
